src/runtime-rs/crates/resource/src/cpu_mem/mem.rs

```diff
@@ -0,0 +1,187 @@
+            for l in &r.hugepage_limits {
+                mem_sb += l.limit;
+            }
+
+            if let Some(memory) = &r.memory {
+                let current_limit = match memory.limit {
+                    Some(limit) => {
+                        mem_sb += limit as u64;
+                        info!(sl!(), "memory sb: {}, memory limit: {}", mem_sb, limit);
+                        limit
+                    }
+                    None => 0,
+                };
+
+                // add swap
+                if let Some(swappiness) = memory.swappiness {
+                    if swappiness > 0 && use_guest_swap {
+                        match memory.swap {
+                            Some(swap) => {
+                                if swap > current_limit {
+                                    swap_sb = swap.saturating_sub(current_limit);
+                                }
+                            }
+                            None => {
+                                if current_limit == 0 {
+                                    need_pod_swap = true;
+                                } else {
+                                    swap_sb += current_limit;
+                                }
+                            }
+                        };
+                    }
+                } // end of if let Some(swappiness)
+            } // end of if let Some(memory)
+        }
+        Ok((mem_sb, need_pod_swap, swap_sb))
+    }
+
+    // update container_cpu_resources field
+    async fn update_container_mem_resources(
+        &self,
+        cid: &str,
+        linux_resources: Option<&LinuxResources>,
+        op: ResourceUpdateOp,
+    ) -> Result<()> {
+        if let Some(r) = linux_resources {
+            let mut resources = self.container_mem_resources.write().await;
+            match op {
+                ResourceUpdateOp::Add | ResourceUpdateOp::Update => {
+                    resources.insert(cid.to_owned(), r.clone());
+                }
+                ResourceUpdateOp::Del => {
+                    resources.remove(cid);
+                }
+            }
+        }
+        Ok(())
+    }
+
+    async fn do_update_mem_resource(
+        &self,
+        new_mem: u32,
+        _swap_sz_byte: i64,
+        hypervisor: &dyn Hypervisor,
+    ) -> Result<u32> {
+        let old_mem = self.current_mem().await;
+        info!(sl!(), "requesting vmm to update memory to {:?}", new_mem);
+        let (new_memory, mem_config) = hypervisor
+            .resize_memory(new_mem, old_mem)
+            .await
+            .context("resize memory")?;
+
+        if hypervisor
+            .capabilities()
+            .await?
+            .is_mem_hotplug_probe_supported()
+            && mem_config.addr != 0
+        {
+            // notify the guest kernel about memory hot-add event, before onlining them
+            // TODO support MemHotplugByProbeRequest
+        }
```

Comment:
Why it's impelementaion with nothing ?

Reply:
Dragonball does need this, so I leave it for other hypervisors to implement it

Reply:
I think we should delete these lines for now, and support this request when we support the hotplug capabilities of other hypervisors...

Reply:
fixed